experimental/personnames/src/api.rs

```diff
@@ -0,0 +1,161 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use std::mem::Discriminant;
+
+use icu_locid::Locale;
+
+/// Trait for providing person name data.
+pub trait PersonName {
+    /// Return the name locale of person name.
+    fn name_locale(&self) -> Option<&Locale>;
+
+    /// return the preferred order of person name.
+    fn preferred_order(&self) -> Option<&PreferredOrder>;
+
+    /// return the value of the given field name, it *must* match the name field.
+    fn get(&self, field: &NameField) -> Option<&str>;
+
+    /// Return all available name field.
+    fn available_name_fields(&self) -> Vec<&NameField>;
+
+    /// Returns true if the provided field name is available.
+    fn has_name_field(&self, lookup_name_field: Discriminant<NameField>) -> bool;
+
+    /// Returns true if person have the name field matching the type and modifier.
+    fn has_name_field_with_modifier(&self, lookup_name_field: &NameField) -> bool;
+}
+
+///
+/// Error handling for the person name formatter.
+///
+#[derive(Debug)]
+pub enum PersonNamesFormatterError {
+    ParseError(String),
+    InvalidPersonName,
+}
+
+/// Field Modifiers.
+///
+/// https://www.unicode.org/reports/tr35/tr35-personNames.html#modifiers
+#[repr(u32)]
+pub enum FieldModifier {
+    Informal = 1 << 0,
+    Prefix = 1 << 1,
+    Core = 1 << 2,
+    AllCaps = 1 << 3,
+    InitialCap = 1 << 4,
+    Initial = 1 << 5,
+    Monogram = 1 << 6,
+}
+
+/// Field Modifiers Mask. (must be the same as FieldModifier repr)
+pub type FieldModifierMask = u32;
```

Comment:
Suggestion: make this a newtype wrapper around the u32 and then you can enforce that it is always valid.

```rust
pub struct FieldModifierSet(u32);

pub enum FieldCapsStyle {
    Auto,
    AllCaps,
    InitialCap,
}

pub enum FieldPart {
    Full,
    Core,
    Prefix,
}

pub enum FieldLength {
    Auto,
    Initial,
    Monogram,
}

pub enum FieldFormality {
    Auto,
    Informal,
}

impl FieldModifierSet {
    pub fn set_caps(&mut self, style: FieldCapsStyle) {
        self = self & !(FieldModifier::AllCaps | FieldModifier::InitialCap) | match style {
            FieldCapsStyle::AllCaps => FieldModifier::AllCaps,
            FieldCapsStyle::InitialCap => FieldModifier::InitialCap,
        };
    }
    // ...
}
```

Reply:
Done.

I don't think you can hide the constructor of a tuple so i changed it into a private named property.
Kept the FieldModifier enum with some twist of Into implementations.

FieldModifierSet only function is ::new and is unmodifiable. (create a new one, but don't change current instance)